// typescript-sdk/src/sdk/change-country.ts
import {ComponentAbstract} from "../common/abstract/component.abstract";
import {ICart, IResponseAuth, ISdkConfig, ISdkSettings} from "../common/interfaces";
import {fromEvent, Observable} from "rxjs";

let tingle = require('tingle.js');

const widthPage = window.outerWidth;
const heightPage = window.outerHeight;

export class ChangeCountry implements ComponentAbstract {
    rawHtml: string;

    private authData: IResponseAuth;
    private cartData: ICart;
    private settings: ISdkSettings;
    private apiConfig: ISdkConfig;

    /** variables de tipo HTML */
    private skybox_change_country: HTMLDivElement;
    private skx_banner_image_account: HTMLElement;
    private skx_position_option_country: HTMLElement;
    private $skx_banner_image_account: Observable<Event>;
    private $skx_position_option_country: Observable<Event>;

    constructor(authData: IResponseAuth, cartData: ICart, settings: ISdkSettings, apiConfig: ISdkConfig) {
        this.authData = authData;
        this.cartData = cartData;
        this.settings = settings;
        this.apiConfig = apiConfig;
        this.loadAssets();
    }

    /** Método que hace un innerHTML a un elemento especifico */
    private drawChangeCountry() {
        /** @ToDo los identificados (nombres, clases, etc) deberian estar en constantes mapeadas, evaluar*/
        this.skybox_change_country = <HTMLDivElement>document.getElementById('skybox-checkout-change-country');
        const content = this.authData.Data.Templates.BarHtmlTemplate;
        this.rawHtml = content;
        this.skybox_change_country.innerHTML = content;
        this.interactionChangeCountry();
    }

    /** Método que implementa las interacciones como el `click` en un elemento espécifico */
    private interactionChangeCountry() {
        this.skx_banner_image_account = <HTMLElement>document.getElementsByClassName('skx_banner_image_account')[0];
        this.skx_position_option_country = <HTMLElement>document.getElementsByClassName('skx_position_option_country')[0];
        if (this.skx_banner_image_account) {
            this.$skx_banner_image_account = fromEvent(this.skx_banner_image_account, 'click');
            this.bannerImageAcountReactive();
        }
        if (this.skx_position_option_country) {
            this.$skx_position_option_country = fromEvent(this.skx_position_option_country, 'click');
            this.positionOptionCountryReactive();
        }
    }

    /** Método que escucha un evento especifico segun una variable reactiva*/
    private bannerImageAcountReactive() {
        this.$skx_banner_image_account.subscribe((e: MouseEvent) => {
            const merchant = this.settings.MERCHANTCODE;
            const cartId = this.authData.Data.CART_SKY.Cart.Id;
            e.preventDefault();
            e.stopPropagation();
            const url = `${this.settings.SKYBOX_URL}APILoginCustomer.aspx?${this.cartData.Data.Cart.DataUrl}&merchant=${merchant}
            &idCart=${cartId}&Reload=1&uri=${this.settings.STORE_URL}?LoadFrame=1`
            const widthpopup = (widthPage - 50).toString();
            const heightpopup = ((heightPage < 800) ? heightPage - 50 : 800).toString();
            this.showPopUp(url, widthpopup + 'px', heightpopup + 'px', 'size-account')
        })
    }

    /** Método que escucha un evento especifico segun una variable reactiva*/
    private positionOptionCountryReactive() {
        this.$skx_position_option_country.subscribe((e: MouseEvent) => {
            e.preventDefault();
            e.stopPropagation();
            const return_url = document.URL.replace('#', '');
            const url = `${this.settings.SKYBOX_URL}Webforms/PublicSite/ReSync.aspx?${this.cartData.Data.Cart.DataUrl}&process_url=${return_url}`;
            var heightpopup = '640px';
            if (/Mobi/.test(navigator.userAgent)) {
                heightpopup = '440px';
            }
            this.showPopUp(url, '540px', heightpopup, 'size-country');
        });
    }

    /**
     * Método para abrir un modal
     * @param {string} url dirección URL del IFRAME
     * @param {string}  w   width
     * @param {string}  h   height
     * @param {string}  className clase CSS
     * */
    private showPopUp(url: string, w: string, h: string, className: string) {
        var modal = new tingle.modal({
            footer: false,
            stickyFooter: false,
            closeMethods: ['overlay', 'button', 'escape'],
            closeLabel: "Close",
            cssClass: ['skybox-modal-content', className],
            onClose: function () {
                top.location.reload();
            }
        });

        modal.setContent(
            '<div class="skybox-iframe-content">' +
            '<iframe width="100%" height="' + h + '" src="' + url + '" frameborder="0" allowfullscreen></iframe>' +
            '</div>'
        );
        modal.open();
    }

    /**
     * setea en el elemento `head` un tag link css
     * */
    private loadAssets() {
        const merchant = this.settings.MERCHANTCODE;
        let head = document.getElementsByTagName('head')[0];
        const release_date = Date.now();
        let link = document.createElement('link');
        link.rel = 'stylesheet';
        link.type = 'text/css';
        link.href = this.apiConfig.SKYBOX_CHECKOUT_URL + 'widgets/api-button/css/api-button-international-css.ashx?s=' + merchant + '&ver=' + release_date;
        link.media = 'all';
        head.appendChild(link);
    }

    getRawHtml() {
        return this.rawHtml;
    }

    render() {
        this.drawChangeCountry();
    }
}
