// typescript-sdk/src/sdk/const.ts
import {ISdkSettings} from "../common/interfaces/index";

export const SDK_SETTINGS: ISdkSettings = {
    IDSTORE: "00269",
    MERCHANT: "ZAA3LTAwMDEyLTYzNTI4MzIwNDcwMzc4NTc3LQ96",
    MERCHANTCODE: "dd497c67-c42c-4e65-be7b-d14f8aa77d21",
    CONF: {
        PRODUCT_TYPE_DEFAULT: "77"
    },
    CHECKOUT_PAGE: "/pages/order-summary",
    CHECKOUT_BUTTON_CLASS: "btn skbx-checkout-btn",
    SUCCESSFUL_PAGE: "/pages/checkout-success",
    STORE_HASH: "demo",
    STORE_URL: "https://sbcmerchant.magento.com/",
    MOD_PLATAFORM: "shopify", // magento - bigcommerce
    CHANGE_COUNTRY_TEXT_USA: "USA ($ USD)",
    SKBX_LOADER_NAME: "loaderOrange.gif",
    INTEGRATION: 1
};

/** @ToDo nombre temporal, consultar por otro mas adecuado */
export const ENDPOINTS = {}

export enum STORAGE_KEY {
    SBC_CART_ID = 'SBC_CART_ID',
    AUTH_STORE = 'auth-store',
    LAST_COUNTRY = 'last_country',
    CURRENCY = 'currency',
    CART_PROD_ARR = 'cart_prod_arr'
}

export const CHECKOUT_BUTTON_TEXT = 'INTERNATIONAL CHECKOUT';
