// typescript-sdk/src/sdk/multicalculate.js
"use strict";
exports.__esModule = true;
var const_1 = require("./const");
var GV = require("../common/global_variables");
var utils_1 = require("../common/utils/utils");
var utils_2 = require("./utils");
var SdkMulticalculate = /** @class */ (function () {
    function SdkMulticalculate(settings, sdkService) {
        this.rutaLoaderGif = "https://s3.amazonaws.com/sky-sbc-images/WebApp/SBC/Images/loaders/loaderGris.gif";
        this.elemBtnAdd = '.Sky--btn-add';
        this.classPrices = 'internationalPrice';
        this.authData = GV.store.get(const_1.STORAGE_KEY.AUTH_STORE);
        this.secondsWaitHtmlMulticalculate = 2000;
        this.settings = settings;
        this.sdkUtils = new utils_2.SdkUtils();
        this.sdkSevice = sdkService;
        if (this.settings.SKBX_LOADER_NAME) {
            this.rutaLoaderGif = "https://s3.amazonaws.com/sky-sbc-images/WebApp/SBC/Images/loaders/" + this.settings.SKBX_LOADER_NAME;
        }
    }
    SdkMulticalculate.prototype.init = function () {
        $(this.elemBtnAdd).attr('disabled', true);
        this.sdkUtils.showLoaders(this.rutaLoaderGif);
        this.multicalculate();
    };
    SdkMulticalculate.prototype.multicalculate = function () {
        var productList = this.proccessMulticalculate();
        if (productList.length)
            this.getMulticalcute(productList);
    };
    SdkMulticalculate.prototype.proccessMulticalculate = function () {
        var _this = this;
        var internationalPrices = Array.from(document.getElementsByClassName(this.classPrices));
        var productList = [];
        internationalPrices.forEach(function (e, index) {
            var elementID = e.getAttribute('id');
            var product = JSON.parse(e.getAttribute('data'));
            var variantID = e.getAttribute('data-variant-id');
            if (elementID && product) {
                var id_shopify_product_variant = String(variantID), title = product.variants[0].name, price = product.variants[0].price / 100, images = product.featured_image, sku = variantID, product_type = product.type;
                var getWeight = _this.sdkUtils.calcWeightAndUnit(product.variants[0].weight);
                if (!id_shopify_product_variant)
                    id_shopify_product_variant = "" + product.id;
                if (_this.settings.CONF.PRODUCT_TYPE_DEFAULT !== '')
                    product_type = _this.settings.CONF.PRODUCT_TYPE_DEFAULT;
                var idObj = id_shopify_product_variant + "__" + _this.sdkUtils.getDate() + _this.authData.Data.CART_SKY.Country.Iso + _this.authData.Data.CART_SKY.Cart.Currency;
                if (product_type !== '') {
                    productList.push({
                        HtmlObjectid: idObj,
                        Sku: sku,
                        Name: title,
                        Category: product_type,
                        Price: price,
                        ImgUrl: images,
                        Weight: getWeight.weight,
                        WeightUnit: getWeight.unit,
                        Optionals: {
                            CustomFields: [
                                product.variants[0].sku
                            ]
                        }
                    });
                }
                else {
                    $(".skbx-loader-" + id_shopify_product_variant).html('<span>Product not available in your country</span>');
                    $(_this.elemBtnAdd).hide();
                }
            }
        });
        return productList;
    };
    SdkMulticalculate.prototype.getMulticalcute = function (producList) {
        var _this = this;
        producList = utils_1.removeDuplicates(producList, 'HtmlObjectid');
        var headerCartId = { 'X-Skybox-Cart-Id': this.authData.Data.Cart.Id };
        this.sdkSevice.getMulticalculate({ ListProducts: producList }, headerCartId).subscribe(function (response) {
            _this.setHTMLMulticalculate(response);
        });
    };
    /** @ToDo falta validar un poco esta función  */
    SdkMulticalculate.prototype.setHTMLMulticalculate = function (response) {
        var _this = this;
        var productsCalculated;
        productsCalculated = response;
        /** agregando la variable `success` con valor `false` a todos los objeteos de la lista */
        productsCalculated.Data.map(function (v) { return v.success = false; });
        productsCalculated.Data.forEach(function (v, i) {
            var productID = v.HtmlObjectId.toString().split('__')[0];
            _this.sdkSevice.getHTMLMulticalculate(v.Url).subscribe(function (response) {
                _this.showPriceMultiC(productID, response);
                if ($(response).is('span.not-available')) {
                    $(_this.elemBtnAdd).hide();
                }
                else {
                    $(response).show();
                    $(_this.elemBtnAdd).attr('disabled', false);
                }
                $('#' + 'skybox-product-price-' + productID).html(response);
                $(".skbx-loader-" + productID).hide();
                productsCalculated.Data[i].success = true;
            }, function (error) {
                $("span[class^='skbx-price']").hide();
            });
        });
    };
    /**
     * @param {string} productID ID del producto que se va cambiar su HTML
     * @param {string} html HTML que se va setear sobre el elemento HTML referenciado por `productID`
     */
    SdkMulticalculate.prototype.showPriceMultiC = function (productId, html) {
        var prefModal = ($(".remodal-is-opened").length > 0) ? ".js-quick-shop " : ""; // only for modal
        $(prefModal + '#skybox-product-price-' + productId).html(html);
        $(prefModal + '#skybox-product-price-' + productId).removeClass('empty-price');
        $(".skbx-loader-" + productId).hide();
    };
    return SdkMulticalculate;
}());
exports.SdkMulticalculate = SdkMulticalculate;
