// typescript-sdk/src/common/utils/http.ts
import {Observable} from 'rxjs';
import 'rxjs/add/observable/dom/ajax';
import {AjaxResponse} from "rxjs/ajax"

export class Http {
    constructor() {

    }

    headers: Object = {
        'Content-Type': 'application/json',
    };

    get(url: string, headers?: Object, responseType = 'json'): Observable<AjaxResponse> {
        return Observable.ajax({
            url: url,
            method: 'GET',
            headers: this.validateWichHeader(headers),
            responseType: responseType
        });
    }

    post(url: string, data: any, headers?: Object): Observable<AjaxResponse> {
        return Observable.ajax({
            url: url,
            method: 'POST',
            headers: this.validateWichHeader(headers),
            body: data
        });
    }

    put(url: string, data: any, headers?: Object): Observable<AjaxResponse> {
        return Observable.ajax({
            url: url,
            method: 'PUT',
            headers: this.validateWichHeader(headers),
            body: data
        });
    }

    patch(url: string, data: any, headers?: Object): Observable<AjaxResponse> {
        return Observable.ajax({
            url: url,
            method: 'PATCH',
            headers: this.validateWichHeader(headers),
            body: data
        });
    }

    delete(url: string, headers?: Object) {
        return Observable.ajax({
            url: url,
            method: 'DELETE',
            headers: this.validateWichHeader(headers),
        });
    }

    setHeaders(headers?: Object) {
        this.headers = headers;
    }

    private validateWichHeader(headers?: Object) {
        return headers ? headers : this.headers ? this.headers : null;
    }
}
