// typescript-sdk/src/common/utils/http.js
"use strict";
exports.__esModule = true;
var rxjs_1 = require("rxjs");
require("rxjs/add/observable/dom/ajax");
var Http = /** @class */ (function () {
    function Http() {
        this.headers = {
            'Content-Type': 'application/json'
        };
    }
    Http.prototype.get = function (url, headers, responseType) {
        if (responseType === void 0) { responseType = 'json'; }
        return rxjs_1.Observable.ajax({
            url: url,
            method: 'GET',
            headers: this.validateWichHeader(headers),
            responseType: responseType
        });
    };
    Http.prototype.post = function (url, data, headers) {
        return rxjs_1.Observable.ajax({
            url: url,
            method: 'POST',
            headers: this.validateWichHeader(headers),
            body: data
        });
    };
    Http.prototype.put = function (url, data, headers) {
        return rxjs_1.Observable.ajax({
            url: url,
            method: 'PUT',
            headers: this.validateWichHeader(headers),
            body: data
        });
    };
    Http.prototype.patch = function (url, data, headers) {
        return rxjs_1.Observable.ajax({
            url: url,
            method: 'PATCH',
            headers: this.validateWichHeader(headers),
            body: data
        });
    };
    Http.prototype["delete"] = function (url, headers) {
        return rxjs_1.Observable.ajax({
            url: url,
            method: 'DELETE',
            headers: this.validateWichHeader(headers)
        });
    };
    Http.prototype.setHeaders = function (headers) {
        this.headers = headers;
    };
    Http.prototype.validateWichHeader = function (headers) {
        return headers ? headers : this.headers ? this.headers : null;
    };
    return Http;
}());
exports.Http = Http;
