// typescript-sdk/src/common/interfaces/sdk.ts
/**
 * Interface de configuración para el SDK
 * */

export interface ISdkSettings {
    IDSTORE: string;
    MERCHANT: string;
    MERCHANTCODE: string;
    CONF: {
        PRODUCT_TYPE_DEFAULT: string
    };
    CHECKOUT_PAGE: string;
    CHECKOUT_BUTTON_CLASS: string;
    SUCCESSFUL_PAGE: string;
    STORE_HASH: string;
    STORE_URL: string;
    MOD_PLATAFORM: string;
    CHANGE_COUNTRY_TEXT_USA: string;
    SKBX_LOADER_NAME: string;
    INTEGRATION: number
    URL?: string;
    SKYBOX_URL?: string;
    IS_API_BETA?: boolean;
    IS_API_PREPROD?: boolean;
}


/**
 * @ToDo algunos nombres de esta interface podrian cambiar
 * por ejemplo
 * `URL` hace referencia a una URL en el cual se va obtener la IP del navegador
 * esta propiedad derrepente podria llamarse `URL_IP`  u otra
 * */
export interface ISdkConfig {
    SKYBOX_CHECKOUT_URL: string;
    SKYBOX_CHECKOUT_API: string;
    SKYBOX_CHECKOUT_AUHT: string;
    SKYBOX_CHECKOUT_AUHT_CART: string;
    SKYBOX_CHECKOUT_CART: string;
    SKYBOX_CHECKOUT_MULTICALCULATE: string;
    SKYBOX_CHECKOUT_API_BASE: string;
    URL_CONTENT: string;
    URL_TIENDAWEB: string;
    URL?: string;
    SKYBOX_URL?: string;
}

/**
 * Interface de respuesta de la API para obtener la IP y información extra
 * del navegador
 * */
export interface IIPData {
    Merchant: {
        Id: string;
        Key: string
    };
    CartId: string;
    Customer: {
        Ip: {
            Local: string;
            Remote: string;
            Proxy: string
        };
        Browser: {
            Agent: string;
            Languages: string
        }
    }
}

/** Interface de respuesta del Api de autenticación */
export interface IResponseAuth {
    Data: {
        Token: string;
        Cart: {
            Id: string;
            DataUrl: string;
        };
        LocationAllow: number;
        IntegrationType: number;
        Boton: {
            Img: string;
            Url: string;
        };
        Templates: {
            BarHtmlTemplate: string;
        };
        CART_SKY?: ICartSky
    }
}

/**
 * Interface del detalle de un Carrito
 * */
export interface ICart {
    Data: {
        ShowLogin: boolean;
        LanguageId: number;
        LocationAllow: number;
        DisplayLogin: string;
        UserTemporal: number;
        StoreCode: string;
        IntegrationType: number;
        Cart: ICartDetail;
        Country: ICountryDetail;
        Template: {
            BarHtmlTemplate: string;
            BarHtmlTemplate2: string;
            BarHtmlTemplate3: string;
            BarHtmlTemplate4: string;
            CssVersion: number
        };
    }
}

export interface ICartDetail {
    Id: string;
    Count: number;
    Currency: string;
    Symbol: string;
    DataUrl: string;
    Height: number;
    Items: ICardDetailItem[]
}

export interface ICardDetailItem {
    Code: string;
    Id: number;
    IsRestricted: boolean;
    Name: string;
    Price: number;
    PriceUSD: number;
    Quantity: number;
    SubTotal: number;
    SubTotalUSD: number;
    HtmlObjectId?: string;
    Product?: { sku: string }; // Only magento
    Errors?: string; // Only magento
}

export interface ICountryDetail {
    Iso: string;
    Name: string;
    Region: string;
    CityId: number;
    City: string;
    Flag: string;
    HascCity: string
}

export interface ICartSky {
    Cart: ICartDetail;
    Country: ICountryDetail;
}

export interface IProduct {
    HtmlObjectid: string;
    Sku: string;
    Name: string;
    Category: string;
    Price: number;
    ImgUrl: string;
    Weight: string;
    WeightUnit: string;
    Optionals?: {
        CustomFields: string[]
    };
}


/** Interfaz de envio al servicio de multicalculate */
export interface IBodyMulticalculate {
    ListProducts: IProduct[];
}


/** Interface respuesta de la llamada del servicio de Multicalculate */
export interface IResponseMulticaculate {
    Data: {
        HtmlObjectId: string;
        Url: string;
        success?: boolean;
    }[]
}

/** Interface de la información de un producto proveniente de Shopify */

export interface IShopifyProduct {
    id: number;
    title: string;
    handle: string;
    description: string;
    published_at: string;
    created_at: string;
    vendor: string;
    type: string;
    tags: string[];
    price: number;
    price_min: number;
    price_max: number;
    available: boolean;
    price_varies: boolean;
    compare_at_price: number;
    compare_at_price_min: number;
    compare_at_price_max: number;
    compare_at_price_varies: boolean;
    variants: IShopifyVariant[];
    images: string[];
    featured_image: string;
    options: string[];
    content: string
}

/** Interface de la información sobre la variante de un producto de Shopify*/

export interface IShopifyVariant {
    id: number;
    title: string;
    option1: string;
    option2: string;
    option3: string;
    sku: string;
    requires_shipping: boolean;
    taxable: boolean;
    featured_image: string;
    available: boolean;
    name: string;
    public_title: string;
    options: string[];
    price: number;
    weight: number;
    compare_at_price: number;
    inventory_management: string;
    barcode: string
}

export interface IShopifyCart {
    token: string;
    /** @ToDo no se que tipo de parametro pueda ser*/
    note: any;
    attributes: {
        custom_language: string
    };
    original_total_price: number;
    total_price: number;
    total_discount: number;
    total_weight: number;
    item_count: number;
    items: IShopifyCartItem[];
    requires_shipping: boolean;
    currency: string
}

export interface IShopifyCartItem {
    id: number;
    /** @ToDo no se que tipo de parametro pueda ser*/
    properties: any;
    quantity: number;
    variant_id: number;
    key: string;
    title: string;
    price: number;
    original_price: number;
    discounted_price: number;
    line_price: number;
    original_line_price: number;
    total_discount: number;
    discounts: number[];
    sku: string;
    grams: number;
    vendor: string;
    taxable: boolean;
    product_id: number;
    gift_card: boolean;
    url: string;
    image: string;
    handle: string;
    requires_shipping: boolean;
    product_type: string;
    product_title: string;
    product_description: string;
    variant_title: string;
    variant_options: string[];
    Code?: string;
    Quantity?: string;
}

/** @ToDo el nombre podria cambiar; consultarlo */
export interface IConcept {
    Data: {
        Product: {
            Local: IProductPriceDetail;
            Usd: IProductPriceDetail;
        };
        Concepts: IConceptDetail[];
        Adjusts: any[]
    }
}

/** @ToDo el nombre podria cambiar; consultarlo */
export interface IConceptDetail {
    Id: number;
    Name: string;
    Price: string;
    PriceUsd: string;
    Currency: string;
    Visible: string;
}

/** @ToDo el nombre podria cambiar; consultarlo */
export interface IProductPriceDetail {
    Total: string;
    Price: string;
    Customs: string;
    Shipping: string;
    Insurance: string;
    Taxes: string;
    Duties: string;
    Handling: string;
    Clearence: string;
    Others: string;
    Adjustment: string;
}

export interface IMagentoDeleteBody {
    idVariant: string;
    Errors: string;
}
