// typescript-sdk/dist/src/sdk/change-country.js
"use strict";
exports.__esModule = true;
var rxjs_1 = require("rxjs");
var tingle = require('tingle.js');
var widthPage = window.outerWidth;
var heightPage = window.outerHeight;
var ChangeCountry = (function () {
    function ChangeCountry(authData, cartData, settings, apiConfig) {
        this.authData = authData;
        this.cartData = cartData;
        this.settings = settings;
        this.apiConfig = apiConfig;
        this.loadAssets();
    }
    ChangeCountry.prototype.drawChangeCountry = function () {
        this.skybox_change_country = document.getElementById('skybox-checkout-change-country');
        var content = this.authData.Data.Templates.BarHtmlTemplate;
        this.rawHtml = content;
        this.skybox_change_country.innerHTML = content;
        this.interactionChangeCountry();
    };
    ChangeCountry.prototype.interactionChangeCountry = function () {
        this.skx_banner_image_account = document.getElementsByClassName('skx_banner_image_account')[0];
        this.skx_position_option_country = document.getElementsByClassName('skx_position_option_country')[0];
        if (this.skx_banner_image_account) {
            this.$skx_banner_image_account = rxjs_1.fromEvent(this.skx_banner_image_account, 'click');
            this.bannerImageAcountReactive();
        }
        if (this.skx_position_option_country) {
            this.$skx_position_option_country = rxjs_1.fromEvent(this.skx_position_option_country, 'click');
            this.positionOptionCountryReactive();
        }
    };
    ChangeCountry.prototype.bannerImageAcountReactive = function () {
        var _this = this;
        this.$skx_banner_image_account.subscribe(function (e) {
            var merchant = _this.settings.MERCHANTCODE;
            var cartId = _this.authData.Data.CART_SKY.Cart.Id;
            e.preventDefault();
            e.stopPropagation();
            var url = _this.settings.SKYBOX_URL + "APILoginCustomer.aspx?" + _this.cartData.Data.Cart.DataUrl + "&merchant=" + merchant + "\n            &idCart=" + cartId + "&Reload=1&uri=" + _this.settings.STORE_URL + "?LoadFrame=1";
            var widthpopup = (widthPage - 50).toString();
            var heightpopup = ((heightPage < 800) ? heightPage - 50 : 800).toString();
            _this.showPopUp(url, widthpopup + 'px', heightpopup + 'px', 'size-account');
        });
    };
    ChangeCountry.prototype.positionOptionCountryReactive = function () {
        var _this = this;
        this.$skx_position_option_country.subscribe(function (e) {
            e.preventDefault();
            e.stopPropagation();
            var return_url = document.URL.replace('#', '');
            var url = _this.settings.SKYBOX_URL + "Webforms/PublicSite/ReSync.aspx?" + _this.cartData.Data.Cart.DataUrl + "&process_url=" + return_url;
            var heightpopup = '640px';
            if (/Mobi/.test(navigator.userAgent)) {
                heightpopup = '440px';
            }
            _this.showPopUp(url, '540px', heightpopup, 'size-country');
        });
    };
    ChangeCountry.prototype.showPopUp = function (url, w, h, className) {
        var modal = new tingle.modal({
            footer: false,
            stickyFooter: false,
            closeMethods: ['overlay', 'button', 'escape'],
            closeLabel: "Close",
            cssClass: ['skybox-modal-content', className],
            onClose: function () {
                top.location.reload();
            }
        });
        modal.setContent('<div class="skybox-iframe-content">' +
            '<iframe width="100%" height="' + h + '" src="' + url + '" frameborder="0" allowfullscreen></iframe>' +
            '</div>');
        modal.open();
    };
    ChangeCountry.prototype.loadAssets = function () {
        var merchant = this.settings.MERCHANTCODE;
        var head = document.getElementsByTagName('head')[0];
        var release_date = Date.now();
        var link = document.createElement('link');
        link.rel = 'stylesheet';
        link.type = 'text/css';
        link.href = this.apiConfig.SKYBOX_CHECKOUT_URL + 'widgets/api-button/css/api-button-international-css.ashx?s=' + merchant + '&ver=' + release_date;
        link.media = 'all';
        head.appendChild(link);
    };
    ChangeCountry.prototype.getRawHtml = function () {
        return this.rawHtml;
    };
    ChangeCountry.prototype.render = function () {
        this.drawChangeCountry();
    };
    return ChangeCountry;
}());
exports.ChangeCountry = ChangeCountry;
