// typescript-sdk/dist/src/sdk/const.js
"use strict";
exports.__esModule = true;
exports.SDK_SETTINGS = {
    IDSTORE: "00269",
    MERCHANT: "ZAA3LTAwMDEyLTYzNTI4MzIwNDcwMzc4NTc3LQ96",
    MERCHANTCODE: "dd497c67-c42c-4e65-be7b-d14f8aa77d21",
    CONF: {
        PRODUCT_TYPE_DEFAULT: "77"
    },
    CHECKOUT_PAGE: "/pages/order-summary",
    CHECKOUT_BUTTON_CLASS: "btn skbx-checkout-btn",
    SUCCESSFUL_PAGE: "/pages/checkout-success",
    STORE_HASH: "demo",
    STORE_URL: "https://sbcmerchant.magento.com/",
    MOD_PLATAFORM: "shopify",
    CHANGE_COUNTRY_TEXT_USA: "USA ($ USD)",
    SKBX_LOADER_NAME: "loaderOrange.gif",
    INTEGRATION: 1
};
exports.ENDPOINTS = {};
var STORAGE_KEY;
(function (STORAGE_KEY) {
    STORAGE_KEY["SBC_CART_ID"] = "SBC_CART_ID";
    STORAGE_KEY["AUTH_STORE"] = "auth-store";
    STORAGE_KEY["LAST_COUNTRY"] = "last_country";
    STORAGE_KEY["CURRENCY"] = "currency";
    STORAGE_KEY["CART_PROD_ARR"] = "cart_prod_arr";
})(STORAGE_KEY = exports.STORAGE_KEY || (exports.STORAGE_KEY = {}));
exports.CHECKOUT_BUTTON_TEXT = 'INTERNATIONAL CHECKOUT';
